sdr_soapy.c

```diff
@@ -0,0 +1,374 @@
+    SoapySDRStream *stream;
+    char* hw_key;
+    bool dev_sdrplay;
+
+    iq_convert_fn converter;
+    struct converter_state *converter_state;
+
+    size_t channel;
+    const char* antenna;
+    bool enable_agc;
+} SOAPY;
+
+//
+// =============================== SoapySDR handling ==========================
+//
+
+void soapyInitConfig()
+{
+    SOAPY.dev = NULL;
+    SOAPY.stream = NULL;
+    SOAPY.hw_key = NULL;
+    SOAPY.dev_sdrplay = false;
+    SOAPY.converter = NULL;
+    SOAPY.converter_state = NULL;
+    SOAPY.channel = 0;
+    SOAPY.antenna = NULL;
+    SOAPY.enable_agc = false;
+}
+
+void soapyShowHelp()
+{
+    printf("      SoapySDR-specific options (use with --device-type soapy)\n");
+    printf("\n");
+    printf("--device <string>        select/configure device\n");
+    printf("--channel <num>          select channel if device supports multiple channels (default: 0)\n");
+    printf("--antenna <string>       select antenna (default depends on device)\n");
+    printf("--enable-agc             enable Automatic Gain Control if supported by device\n");
+
+    printf("\n");
+}
+
+bool soapyHandleOption(int argc, char **argv, int *jptr)
+{
+    int j = *jptr;
+    bool more = (j +1  < argc);
+
+    if (!strcmp(argv[j], "--channel")) {
+        SOAPY.channel = atoi(argv[++j]);
+    } else if (!strcmp(argv[j], "--antenna") && more) {
+        SOAPY.antenna = strdup(argv[++j]);
+    } else if (!strcmp(argv[j], "--enable-agc") && more) {
+        SOAPY.enable_agc = true;
+    } else {
+        return false;
+    }
+
+    *jptr = j;
+    return true;
+}
+
+bool soapyOpen(void)
+{
+    size_t length = 0;
+    SoapySDRKwargs *results = SoapySDRDevice_enumerate(NULL, &length);
+    for (size_t i = 0; i < length; i++)
+    {
+        printf("Found device #%d: ", (int)i);
+        for (size_t j = 0; j < results[i].size; j++)
+        {
+            printf("%s=%s", results[i].keys[j], results[i].vals[j]);
+            if (j+1 < results[i].size) printf(", ");
+            if (!strncmp(results[i].keys[j], "driver", 6)) {
+                if (!strncmp(results[i].vals[j], "sdrplay", 7)) {
+                    SOAPY.dev_sdrplay = 1;
+                }
+            }
+        }
+        printf("\n");
+    }
+    SoapySDRKwargsList_clear(results, length);
+
+    SOAPY.dev = SoapySDRDevice_makeStrArgs(Modes.dev_name);
+    if (!SOAPY.dev) {
+        fprintf(stderr, "soapy: failed to create device: %s\n", SoapySDRDevice_lastError());
+        return false;
+    }
+
+    SoapySDRKwargs result = SoapySDRDevice_getHardwareInfo(SOAPY.dev);
+    if (result.size > 0) {
+        for (size_t i = 0; i < result.size; ++i) {
+            printf("%s=%s", result.keys[i], result.vals[i]);
+            if (i+1 < result.size) printf(", ");
+        }
+        printf("\n");
+    }
+    SoapySDRKwargs_clear(&result);
+
+    char* hw_key = SoapySDRDevice_getHardwareKey(SOAPY.dev);
+    if (hw_key) {
+        printf("soapy: hardware key is %s\n", hw_key);
```

Comment:
What's this actually used for? I couldn't find references to hw_key elsewhere.

Reply:
I was anticipating needing it for RTLSDR or RSPduo specific behavior but you're right, so far haven't needed it, so I went ahead and removed it,  (Although I am still wondering about setting gain for RTLSDR.  I'll refer back to that in my upcoming reply to your gain API comment.)